javaslang/src/main/java/javaslang/collection/Arrays.java

```diff
@@ -0,0 +1,80 @@
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2016 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+/**
+ * Internal class, containing Java array manipulation helpers.
+ * Many arrays are represented as simple Objects, to avoid casts at the client site, and to work with all types of arrays, including primitive ones.
+ *
+ * @author Pap Lőrinc
+ * @since 2.1.0
+ */
+final class Arrays { // TODO reuse these in `Array` also
+    private static final Object[] EMPTY = {};
+    @SuppressWarnings("unchecked")
+    static <T> T[] emptyArray() { return (T[]) EMPTY; }
+
+    /** Create a single element array */
+    static Object[] asArray(Object element) {
+        return new Object[] { element };
+    }
+
+    /** System.arrayCopy with same source and destination */
+    static Object[] copyRange(Object array, int from, int to) {
+        final int length = to - from;
+        return arrayCopy(length, array, from, 0, length);
+    }
+
+    /** clone the source and set the value at the given position */
+    static Object[] copyUpdate(Object arrayObject, int index, Object element) {
+        final Object[] array = (Object[]) arrayObject;
+        final Object[] copy = copy(array, index + 1);
+        copy[index] = element;
+        return copy;
+    }
+
+    static Object[] copy(Object arrayObject, int minLength) {
+        final Object[] array = (Object[]) arrayObject;
+        final int arrayLength = (array == null) ? 0 : array.length;
+        final int length = Math.max(arrayLength, minLength);
+        if (length == arrayLength) {
+            return array.clone();
+        } else {
+            return arrayCopy(length, array, 0, 0, arrayLength);
+        }
+    }
+
+    /** clone the source and keep everything after the index (pre-padding the values with null) */
+    static Object[] copyDrop(Object arrayObject, int index) {
+        final Object[] array = (Object[]) arrayObject;
+        final int length = array.length;
+        return arrayCopy(length, array, index, index, length - index);
+    }
+
+    /** clone the source and keep everything before the index */
+    static Object[] copyTake(Object arrayObject, int lastIndex) {
+        final Object[] array = (Object[]) arrayObject;
+        return copyRange(array, 0, lastIndex + 1);
+    }
+
+    /** for performance reasons the array allocation and the System.arraycopy must be next to each other - even an if (size > 0) will make it slower */
+    static Object[] arrayCopy(int arraySize, Object source, int sourceFrom, int destinationFrom, int size) {
+        if (size == 0) {
+            return new Object[arraySize];
+        } else {
+            final Object[] result = new Object[arraySize];
+            System.arraycopy(source, sourceFrom, result, destinationFrom, size);
```

Comment:
this was a weird one ... if I put an `if` statement in front of the `arraycopy`, the performance will be halved (i.e. `if (size > 0)`, which should only simplify things...)